database/seed.js

```diff
@@ -31,10 +31,11 @@ Promise.all(colours.reduce((terms, colour) => [
       .map((_, i) => i + productIdStart)
       .map(productId => {
         const animal = animals[Math.floor(Math.random() * animals.length)];
+        const csi = Math.floor(Math.random() * 8); // Colour Slice Index
         return {
           productId,
           productName: `${faker.name.firstName()} the ${animal}`,
-          colours: colours.map(colourName => {
+          colours: [...colours.slice(0, csi), ...colours.slice(csi + 1)].map(colourName => {
```

Comment:
I see where csi comes from, but I don't understand why Math.random() is multiplied by 8 (yielding a number from 0 to 7, correct?). The colours array has (currently) and length of 4. Without running this, it's hard for me to understand what it's doing. My intuition is that it's either grabbing one of the colours in the array, or else grabbing nothing, depending on the value of the csi? Is there maybe a more overt way to accomplish this? I also wonder if changing the number of colours in the array would make this stop working as intended. I admit freely, I may be misunderstanding the workings of this entirely. If so, maybe disregard.

Reply:
The idea here is that I want the seeded database to have products where 50% of them have three colours, and the rest have four. Of the ones with three, there should be an even probability that any one of the four colours is the one that's missing.

This was my way of accomplishing the assignment requirement of a varied number of colours per product without adding lots of code to do the job. Generate four colours, delete one if `csi` is within range, or do nothing if it's out of range.

You could also think of this as an alternative to `Array.prototype.splice` that doesn't modify the original array. In this case, because that original array (`colours` will be used once per productId).

(I'll make a small change to make the intention of that `8` more clear...)